### DISCUSSION TITLE
Unable to cross compile OpenDDS for linux using CMake

### DISCUSSION THREAD

**Discussion by pavanscichip** [OpenDDS/OpenDDS#4596] [Q&A]
Hi I am trying to build OpenDDS for "aarch64" in linux and below mentioned the configuration for the same

Target: aarch64 (variscite imx8mp) | linux | yocto
Host: x86_64 | Ubuntu 20.04 
OpenDDS version: 3.14
steps I did for cross compile are below
1. source toolchain environment setup
    **$ source /hdd/phytec-sdk/environment-setup-aarch64-phytec-linux**
2. configure OpenDDS source for cross compile
    **$ ./configure --no-tests --target=linux-cross --target-compiler=$(which aarch64-fslc-linux-g++)**
3. build OpenDDS
     **$ make**
below shows the complete log for the same

**raibos@scichip-01:~/temp/DDS-3.14$ source /hdd/variscite-sdk/environment-setup-armv8a-fslc-linux** 
**raibos@scichip-01:~/temp/DDS-3.14$ ./configure --no-tests --target=linux-cross --target-compiler=$(which aarch64-fslc-linux-g++)**
Downloading ACE+TAO 2.2a with latest patches
Extracting archive ACE+TAO-2.2a_with_latest_patches_NO_makefiles.tar.gz
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = "en_IN:en",
	LC_ALL = (unset),
	LANG = "en_IN"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").
Searching /home/raibos/temp/DDS-3.14 for files...
Found 33232 files and directories in 0 seconds.
Creating /home/raibos/temp/DDS-3.14/build/host
Finished in /home/raibos/temp/DDS-3.14/build/host
Creating /home/raibos/temp/DDS-3.14/build/target
Finished in /home/raibos/temp/DDS-3.14/build/target
Total time: 3 seconds.
Cross-compile configuring host
Running MPC to generate makefiles.
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = "en_IN:en",
	LC_ALL = (unset),
	LANG = "en_IN"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").
MPC_ROOT was set to /home/raibos/temp/DDS-3.14/ACE_wrappers/MPC.
Using .../DDS-3.14/ACE_wrappers/bin/MakeProjectCreator/config/MPC.cfg
Generating 'gnuace' output using build/host/host_tools.mwc
Skipping TAO_IDL_GEN (tao_idl_fe.mpc); it requires tao_idl_fe_gen.
Skipping idl2jni_codegen (idl2jni_codegen.mpc); it requires java.
Generation Time: 1s
Cross-compile configuring target
Running MPC to generate makefiles.
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = "en_IN:en",
	LC_ALL = (unset),
	LANG = "en_IN"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").
MPC_ROOT was set to /home/raibos/temp/DDS-3.14/ACE_wrappers/MPC.
Using .../DDS-3.14/ACE_wrappers/bin/MakeProjectCreator/config/MPC.cfg
Generating 'gnuace' output using build/target/DDS_TAOv2.mwc
Skipping ACE_XML_Utils (XML.mpc); it requires xerces.
Skipping gperf (gperf.mpc); it avoids cross_compile.
Skipping TAO_IDL_GEN (tao_idl_fe.mpc); it requires tao_idl_fe_gen.
Skipping OpenDDS_Security (DdsSecurity.mpc); it requires xerces.
Skipping OpenDDS_XML_QOS_XSC_Generation (qos_xml_handler.mpc); it requires xsc.
Skipping OpenDDS_QOS_XML_XSC_Handler (qos_xml_handler.mpc); it requires xerces.
Skipping OpenDDS_CorbaSeq (CorbaSeq.mpc); it avoids no_opendds_safety_profile.
Skipping opendds_idl (opendds_idl.mpc); it avoids cross_compile.
Skipping OpenDDS_Corba (OpenDDS_Corba.mpc); it avoids no_opendds_safety_profile.
Skipping Monitor_App (Monitor.mpc); it requires qt5.
Skipping dissector (dissector.mpc); it requires wireshark_any.
Skipping ExcelRTD (ExcelRTD.mpc); it requires qt5.
Skipping tao_java (tao_java.mpc); it requires java.
Skipping idl2jni_corba (idl2jni_corba.mpc); it requires java.
Skipping idl2jni_codegen (idl2jni_codegen.mpc); it requires java.
Skipping idl2jni_test_simple (idl2jni_test_simple.mpc); it requires java.
Skipping idl2jni_test_union (idl2jni_test_union.mpc); it requires java.
Skipping idl2jni_runtime (idl2jni_runtime.mpc); it requires java.
Skipping two_idl (two_idl.mpc); it requires java.
Skipping transport_config (transport_config.mpc); it requires java.
Skipping messenger_idl_test (messenger_idl_test.mpc); it requires java.
Skipping publisher_idl_test (publisher_idl_test.mpc); it requires java.
Skipping subscriber_idl_test (subscriber_idl_test.mpc); it requires java.
Skipping java_both_test (java_both_test.mpc); it requires java.
Skipping complex_idl_test (complex_idl_test.mpc); it requires java.
Skipping multirepo_test (multirepo_test.mpc); it requires java.
Skipping builtintopics_test (builtintopics_test.mpc); it requires java.
Skipping hello_java_client (hello_java_client.mpc); it requires java.
Skipping participant_location_test (participant_location_test.mpc); it requires java.
Skipping zerocopy_java_test (zerocopy.mpc); it requires java.
Skipping dcps_java (dcps_java.mpc); it requires java.
Generation Time: 29s
Completed configuring OpenDDS, next run 'make' to compile 
the host and target builds of OpenDDS and ACE+TAO.
You can use the generated setenv.sh script in the
build/host and build/target dirs to set environment
variables for future shell sessions.
**raibos@scichip-01:~/temp/DDS-3.14$ cd build/target/**
**raibos@scichip-01:~/temp/DDS-3.14/build/target$ make**
make[1]: Entering directory '/home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace'
touch .depend.ACE

GNUmakefile: /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/GNUmakefile.ACE MAKEFLAGS=w

/hdd/variscite-sdk/sysroots/x86_64-fslcsdk-linux/usr/bin/aarch64-fslc-linux/aarch64-fslc-linux-g++ -fvisibility=hidden -fvisibility-inlines-hidden -Wnon-virtual-dtor -ggdb -pthread -fno-strict-aliasing -Wall -W -Wpointer-arith -pipe  -D_GNU_SOURCE   -I/home/raibos/temp/DDS-3.14/build/target/ACE_wrappers -D__ACE_INLINE__ -I.. -DACE_BUILD_DLL  -c -fPIC -o arm64/.shobj/Local_Name_Space.o Local_Name_Space.cpp
In file included from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/config-linux.h:24,
                 from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/config.h:1,
                 from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/config-macros.h:24,
                 from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/config-lite.h:24,
                 from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/ACE_export.h:11,
                 from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/Malloc.h:18,
                 from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/Malloc_T.h:18,
                 from /home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/Local_Name_Space.h:20,
                 from Local_Name_Space.cpp:3:
/home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace/config-linux-common.h:18:10: fatal error: features.h: No such file or directory
   18 | #include <features.h>
      |          ^~~~~~~~~~~~
compilation terminated.
make[1]: *** [/home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/include/makeinclude/rules.local.GNU:190: arm64/.shobj/Local_Name_Space.o] Error 1
make[1]: Leaving directory '/home/raibos/temp/DDS-3.14/build/target/ACE_wrappers/ace'
make: *** [GNUmakefile.dist:109: ACE] Error 2

**Comment by mitza-oci** [comment]
Please see https://opendds.readthedocs.io/en/latest-release/devguide/building/index.html#cross-compiling and https://opendds.readthedocs.io/en/latest-release/devguide/building/index.html#id3 -- have you tried the current release of OpenDDS?  3.14 is old and many improvements to build/cmake/cross-compile have occurred since then.

**Comment by vermaete** [comment]
@raibopavan 
When using Phytec and a by Yocto generated SDK, why not compiling OpenDDS in Yocto itself?  See: http://layers.openembedded.org/layerindex/recipe/111644/

**Reply by pavanscichip** [reply]
Hi Vermaete,

Yes. We can build OpenDDS in yocto itself. I did it for two versions of two different boards
1. Phytec
2. Variscite

after installing the toolchains of the two varieties below are the outputs for building the sample application
```
**#####################PHYTEC######################**

raibos@scichip-01:~/temp/Qt_dds$ source /hdd/phytec-sdk/environment-setup-aarch64-phytec-linux 
raibos@scichip-01:~/temp/Qt_dds$ cd phytec/
raibos@scichip-01:~/temp/Qt_dds/phytec$ cmake ..
-- Toolchain file defaulted to '/hdd/phytec-sdk/sysroots/x86_64-phytecsdk-linux/usr/share/cmake/OEToolchainConfig.cmake'
-- The C compiler identification is GNU 9.2.0
-- The CXX compiler identification is GNU 9.2.0
-- Check for working C compiler: /hdd/phytec-sdk/sysroots/x86_64-phytecsdk-linux/usr/bin/aarch64-phytec-linux/aarch64-phytec-linux-gcc
-- Check for working C compiler: /hdd/phytec-sdk/sysroots/x86_64-phytecsdk-linux/usr/bin/aarch64-phytec-linux/aarch64-phytec-linux-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /hdd/phytec-sdk/sysroots/x86_64-phytecsdk-linux/usr/bin/aarch64-phytec-linux/aarch64-phytec-linux-g++
-- Check for working CXX compiler: /hdd/phytec-sdk/sysroots/x86_64-phytecsdk-linux/usr/bin/aarch64-phytec-linux/aarch64-phytec-linux-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Creating IDL compiler command for /home/raibos/temp/Qt_dds/Publisher/../idl/Topics.idl
-- Creating IDL compiler command for /home/raibos/temp/Qt_dds/Publisher/../idl1/Topics1.idl
-- Creating IDL compiler command for /home/raibos/temp/Qt_dds/Subscriber/../idl/Topics.idl
-- Creating IDL compiler command for /home/raibos/temp/Qt_dds/Subscriber/../idl1/Topics1.idl
-- Configuring done
-- Generating done
-- Build files have been written to: /home/raibos/temp/Qt_dds/phytec
raibos@scichip-01:~/temp/Qt_dds/phytec$ make
Scanning dependencies of target Publisher_autogen
[  2%] Compiling /home/raibos/temp/Qt_dds/Publisher/../idl/Topics.idl
/hdd/phytec/yocto/build/tmp/work/x86_64-nativesdk-phytecsdk-linux/nativesdk-opendds/3.14-r0/git/x86_64-phytecsdk-linux-g++: 2: x86_64-phytecsdk-linux-g++: not found
/hdd/phytec-sdk/sysroots/x86_64-phytecsdk-linux/usr/bin/tao_idl: preprocessor "/hdd/phytec/yocto/build/tmp/work/x86_64-nativesdk-phytecsdk-linux/nativesdk-opendds/3.14-r0/git/x86_64-phytecsdk-linux-g++" returned with an error
make[2]: *** [Publisher/CMakeFiles/Publisher_autogen.dir/build.make:126: Publisher/TopicsC.cpp] Error 1
make[1]: *** [CMakeFiles/Makefile2:125: Publisher/CMakeFiles/Publisher_autogen.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
raibos@scichip-01:~/temp/Qt_dds/phytec$ 


**####################VARISCITE#####################**

raibos@scichip-01:~/temp/Qt_dds$ source /hdd/variscite-sdk/environment-setup-armv8a-fslc-linux 
raibos@scichip-01:~/temp/Qt_dds$ cd variscite/
raibos@scichip-01:~/temp/Qt_dds/variscite$ cmake ..
-- Toolchain file defaulted to '/hdd/variscite-sdk/sysroots/x86_64-fslcsdk-linux/usr/share/cmake/OEToolchainConfig.cmake'
-- The C compiler identification is GNU 11.3.0
-- The CXX compiler identification is GNU 11.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /hdd/variscite-sdk/sysroots/x86_64-fslcsdk-linux/usr/bin/aarch64-fslc-linux/aarch64-fslc-linux-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /hdd/variscite-sdk/sysroots/x86_64-fslcsdk-linux/usr/bin/aarch64-fslc-linux/aarch64-fslc-linux-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
--   Using OpenDDS 3.27.0 at /hdd/variscite-sdk/sysroots/armv8a-fslc-linux/usr/share/dds
--   Using MPC at /hdd/kirkstone/var-fsl-yocto/build_xwayland/tmp/work/armv8a-fslc-linux/opendds/3.27.0-r0/git/ACE_wrappers/MPC
--   Using ACE 7.1.3 at /hdd/variscite-sdk/sysroots/armv8a-fslc-linux/usr/share/ace
--   Using TAO 3.1.3 at /hdd/variscite-sdk/sysroots/armv8a-fslc-linux/usr/share/tao
--   Looking for pthread.h
--   Looking for pthread.h - found
--   Performing Test CMAKE_HAVE_LIBC_PTHREAD
--   Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
--   Found Threads: TRUE  
CMake Error at /hdd/variscite-sdk/sysroots/armv8a-fslc-linux/usr/share/cmake/OpenDDS/import_common.cmake:297 (message):
  Missing required dependencies _OPENDDS_ACE_ACE_GPERF
Call Stack (most recent call first):
  /hdd/variscite-sdk/sysroots/armv8a-fslc-linux/usr/share/cmake/OpenDDS/OpenDDSConfig.cmake:209 (_opendds_found_required_deps)
  Publisher/CMakeLists.txt:14 (find_package)


CMake Error at Publisher/CMakeLists.txt:14 (find_package):
  Found package configuration file:

    /hdd/variscite-sdk/sysroots/armv8a-fslc-linux/usr/share/cmake/OpenDDS/OpenDDSConfig.cmake

  but it set OpenDDS_FOUND to FALSE so package "OpenDDS" is considered to be
  NOT FOUND.


-- Configuring incomplete, errors occurred!
See also "/home/raibos/temp/Qt_dds/variscite/CMakeFiles/CMakeOutput.log".
raibos@scichip-01:~/temp/Qt_dds/variscite$ 
```

Could you please help resolving these errors
Thanks in advance
Pavan M

**Reply by pavanscichip** [reply]
In addition to the above, I have built the Yocto of the following combinations

1. Phytec | Yocto Dunfell | OpenDDS_3.14
2. Variscite | Yocto Dunfell | OpenDDS_3.14
3. Variscite | Yocto Kirkstone | OpenDDS_3.27

But the results are the same as above.

**Reply by mitza-oci** [reply]
Please ask questions about OpenDDS's yocto support in its own repository https://github.com/OpenDDS/meta-opendds

**Reply by jwillemsen** [reply]
We build ACE/TAO/OpenDDS as part of AXCIOMA for Yocto 3.4.4 without problems, but we don't use cmake, but MPC with the ACE cross build support, see https://www.axcioma.org/scoreboard for our build logs

**Reply by iguessthislldo** [reply]
>   Missing required dependencies _OPENDDS_ACE_ACE_GPERF

Is [`OPENDDS_HOST_TOOLS`](https://opendds.readthedocs.io/en/master/devguide/building/cmake.html#cmake-var-OPENDDS_HOST_TOOLS) set?

**Reply by iguessthislldo** [reply]
I would strongly advise against trying to cross compile an application using CMake and OpenDDS 3.14. As Adam mentioned it's quite old, but the CMake support was added for the first time in 3.14 and cross compiling wasn't supported at all. Later versions can be [convinced to do it](https://github.com/OpenDDS/OpenDDS/discussions/4199#discussioncomment-6432409), but the proper support didn't exist until 3.26.

Also maybe the CMake support could see that the host tools are missing for a cross compile and suggest the correct variables to set.
